src/focus-trap-react.js

```diff
@@ -55,24 +60,35 @@ class FocusTrap extends React.Component {
     }
   }
 
-  /** Returns focus to the element that had focus when the trap was activated. */
-  returnFocus() {
-    const { checkCanReturnFocus, onPostDeactivate } =
-      this.props.focusTrapOptions;
+  deactivateTrap() {
+    const { checkCanReturnFocus } = this.tailoredFocusTrapOptions;
+
+    if (this.focusTrap) {
+      // NOTE: we never let the trap return the focus since we do that ourselves
+      this.focusTrap.deactivate({ returnFocus: false });
+    }
 
-    const sendFocus = () => {
-      if (this.previouslyFocusedElement?.focus) {
+    const finishDeactivation = () => {
+      const canReturnFocus =
+        this.previouslyFocusedElement?.focus && this.returnFocusOnDeactivate;
+
+      if (canReturnFocus) {
+        /** Returns focus to the element that had focus when the trap was activated. */
         this.previouslyFocusedElement.focus();
-        if (onPostDeactivate) {
-          onPostDeactivate();
-        }
+      }
+
+      if (this.onPostDeactivate) {
+        this.onPostDeactivate.call(null); // don't call it in context of "this"
       }
     };
 
     if (checkCanReturnFocus) {
-      checkCanReturnFocus(this.previouslyFocusedElement).then(sendFocus);
+      checkCanReturnFocus(this.previouslyFocusedElement).then(
```

Comment:
Instead of just `this.previouslyFocusedElement`, it should look more like https://github.com/focus-trap/focus-trap/blob/master/index.js#L561 so that `setReturnFocus` option is handled properly since it won't be handled by focus-trap now.